### PR TITLE
Get manila api microversion

### PR DESCRIPTION
**Pull request by pospispa**
For https://github.com/gophercloud/gophercloud/issues/340

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

https://github.com/openstack/manila/blob/stable/newton/manila/api/versions.py#L88

### CODE REVIEW

**Comment by pospispa** [comment]
@jtopjian @jrperritt please review

**Comment by jtopjian** [comment]
@pospispa Thanks for working on this! I'll be able to dedicate some time to this over the next few days.

**Comment by pospispa** [comment]
@jtopjian thank you for letting me know.
I'm looking forward to your comments.

**Comment by jrperritt** [comment]
On hold for a decision on #341

**Comment by pospispa** [comment]
@jtopjian @jrperritt this PR was waiting for decision on https://github.com/gophercloud/gophercloud/pull/341
As https://github.com/gophercloud/gophercloud/pull/341 is closed without merge and the https://github.com/gophercloud/gophercloud/pull/355 is merged I updated this PR according to current master.
I also updated https://github.com/gophercloud/gophercloud/pull/346 that is based on this PR, however, is still WIP.

To sum it up, this PR is ready for review. Please, take a look.

**Comment by pospispa** [comment]
@jtopjian please, do you plan to take a look on this PR.

**Comment by pospispa** [comment]
@jtopjian I'm sorry, I don't precisely understand what you mean. Please, do you mean that sharedfilesystems package should use the [func ListVersions](https://github.com/gophercloud/gophercloud/blob/master/openstack/orchestration/v1/apiversions/requests.go#L9) to get the current sharedfilesystems Microversion?

**Comment by jtopjian** [comment]
Yes, kind of. It should use a `ListVersions` function, but that function is probably better suited in a dedicated `sharedfilesystems/apiversions` package (note how it's not `sharedfilesystems/v2/apiversions` since, in theory, any version could/should be able to use this functionality. This differs from how the existing `apiversion` packages are today, but that will most likely change to them being placed into a non-versioned directory).

The reason for the dedicated package is because this functionality has a high probability of re-use. You've implemented microversion detection for `shares`, but what if `sharetypes` then also needs to work with microversions?

Let me know if that clarifies things :)

**Comment by pospispa** [comment]
@jtopjian yes it clarifies things. Now, I understand your original comment better.

**Comment by pospispa** [comment]
@jtopjian please, let me try to explain my use case. A new service client for a particular shared filesystem endpoint is created in [func NewSharedFileSystemV2](https://github.com/gophercloud/gophercloud/blob/master/openstack/client.go#L263). The `func NewSharedFileSystemV2` finds an endpoint that is major version specific, e.g. `/v2/`. I need to extend the `func NewSharedFileSystemV2` in such a way that the func except finding endpoint the [Microversion](https://github.com/gophercloud/gophercloud/blob/master/service_client.go#L30) is also initialized to the [Microversion](https://docs.openstack.org/developer/manila/devref/api_microversion_dev.html) value of the found endpoint.

In previous comments we touched orchestration service API [list versions](https://developer.openstack.org/api-ref/orchestration/v1/index.html?expanded=list-versions-detail) that lists all **MAJOR** API versions of orchestration service and is at [`/`](https://developer.openstack.org/api-ref/orchestration/v1/index.html?expanded=list-versions-detail).
I agree that functions for listing all **MAJOR** API versions of an OpenStack service shall be in `openstack/servicename/apiversions` package.

However, my use case is not listing all major versions of shared filesystems service at [`/`](https://developer.openstack.org/api-ref/shared-file-systems/?expanded=list-all-major-versions-detail), however, I need to get Microversion of a particular endpoint that already has a particular **MAJOR** version in its path, e.g. [`/v2/`](https://developer.openstack.org/api-ref/shared-file-systems/?expanded=show-details-of-specific-api-version-detail). That's why in my opinion the package should be `sharedfilesystems/v2/microversion` because it's **MAJOR** version specific.

Please, do you agree?

**Comment by jtopjian** [comment]
@pospispa The reason I suggested placing the `apiversions` package in a non-versioned directory is because it is possible to get all versions (both major and minor) by querying the root URL of the service.

As an exercise:

```bash
#!/bin/bash

oscurl() {
  token=$(openstack token issue -c id -f value)
  curl -s -H "X-Auth-Token:$token" $1
}

echo "Cinder API versions"
oscurl http://localhost:8776/ | python -mjson.tool

echo "Nova API versions"
oscurl http://localhost:8774/ | python -mjson.tool

echo "Glance API versions"
oscurl http://localhost:9292/ | python -mjson.tool

echo "Manila API versions"
oscurl http://localhost:8786/ | python -mjson.tool

echo "Manila v2 API versions"
oscurl http://localhost:8786/v2 | python -mjson.tool
```

You _should_ get an error on the last call to `oscurl` which is querying Manila's `/v2` endpoint. However, the second-to-last call is querying Manila's root URL which is returning major and minor versions for both v1 and v2.

If you were to put `apiversions` in v2 and if someone wanted to add `apiversions` to v1 (or v3 in the future), wouldn't the only difference in code be changing the number "2"?

**Comment by jtopjian** [comment]
Actually, as soon as I posted, I saw that

```
oscurl http://localhost:8786/v2 | python -mjson.tool
```

was missing a `/` at the end of the URL and it looks like adding the `/` makes a difference:

```
oscurl http://localhost:8786/v2/ | python -mjson.tool
```

That will return all of the v2 results. However, these results are already returned as part of the larger result set when doing:

```
oscurl http://localhost:8786/ | python -mjson.tool
```

So I still think putting `apiversions` in a non-versioned directory is still applicable.

**Comment by pospispa** [comment]
@jtopjian I agree that having `apiversions` in non-versioned directory for querying `http://localhost:8786/`
is the correct way.

However, I need endpoint specific, i.e. major version specific microversion.

I can think of 2 possible ways how to get the endpoint specific microversion.
1. Using `apiversions` package:
a. Query `http://localhost:8786/` to get all available versions `allVersions`(note: this query expects HTTP status code 300).
b. Parse endpoint URL to get it's major version `epMajorVer`, e.g. `v2`.
c. Find `epMajorVer` specific records `epRec` in `allVersions`, choose one record from `epRec` and get the endpoint specific microversion from it.

2. Using `v2/microversion` package:
a. Query `http://localhost:8786/v2/` to get endpoint specific versions `epVersions`(note: this query expects HTTP status code 200).
b. Choose one record from `epVersions` and get the endpoint specific microversion from it.

I prefer the second way (to query `http://localhost:8786/v2/`) because IMHO it's easier.

@jtopjian please, what's your opinion.

I understand that you have concerns that in case `v2/microversion` package is added that there may be code duplication at the time `v3/microversion` package is added or when `apiversions` package is added because extraction of the results is the same and requests are same except URL and expected HTTP status code.

**Comment by pospispa** [comment]
The PR #409 implements the Get microversion request in a better way. That's why I'm closing this PR.
